Automattic/wp-calypso
# Make client/ directory (Calypso app) a Lerna package

Turns the Calypso client (and server) into proper Lerna monorepo package. The top-level `package.json` now contains only dev dependencies for the whole monorepo, and there is a new `client/package.json` that now has the runtime dependencies for Calypso client and server.

The main benefit is that we can now have a single unified `node_modules` tree, installed with one `npm ci` command, and with one global `package-lock.json`. Separate lockfiles and NPM installs in `apps/` were removed. The only part of the monorepo that continues to have its own NPM tree is `tests/e2e`.

The `server/` directory was moved to `client/server/` to make the Calypso app self-contained inside one directory. Many import paths in the `server/` sources needed to be updated and clarified.

The root `package.json` declares the app-like subpackages as `dependencies` to make them part of the `npm i` tree.

Some `devDependencies` were incorrectly declared as `dependencies`, so I moved them to the right place.

**Some areas that are not 100%**
- in `Dockerfile`, I had to disable the separate layer for `npm ci`. If I understand the logic correctly, it first copies a minimal version of the source tree that's needed to run the `npm install`, then installs the `node_modules`, and only then copies and builds the rest? @sirreal it would be nice if we could make it work again
- there are probably more `eslint-plugin-import` errors reported by ESLint now than needed. Can we improve on that?
- the `wp-desktop` is failing. 🙂 

**How to test:**
- test the Full Site Editing and WP.com Block Editor apps -- do they build and run as expected? Cc @Automattic/cylon as I personally have only limited knowledge about these apps.
- does Calypso really work 100%? The hardest part to make right were devdocs.
